### ISSUE TITLE
simple torch.nn.Linear BUG with reproduce code

### ISSUE THREAD

**Issue by Chandler-Bing** [pytorch/pytorch#122778]
### 🐛 Describe the bug

```python
import torch

m = torch.load(f'./gate_linear.pt',map_location='cpu')
print(m)
#Linear(in_features=1024, out_features=2816, bias=False)

x1 = torch.load(f'./x1.pt',map_location='cpu')
x2 = torch.load(f'./x2.pt',map_location='cpu')
print(x1.size(),x2.size())
#torch.Size([1, 128, 1024]) torch.Size([1, 129, 1024])

print(torch.equal(x1,x2[:,:-1,:]))
#True

o1 = m(x1)
o2 = m(x2)

print(torch.equal(
    o1,
    o2[:,:-1,:]
    ))

#False
```
the code is simple enough, as the True result of torch.equal suggest, input x1, x2[:,:-1,:] should be same.
why it become different after a simple Linear model?
I upload my pt files (about 4M size) to reproduce the result, [google drive](https://drive.google.com/file/d/1E1ShH3j6QrgS8wRRqopz8pLQST-0qZ_T/view?usp=sharing)
any thoughts? thank you guys

minimal reproducible code …
```python

import torch
torch.set_printoptions(precision=8)
x = torch.tensor([[[ 0.0451, -0.8093],
                   [-0.3275, -0.5304]]]
                 )
x1 = x[:,:-1,:]
x2 = x
w = torch.tensor(   [[ 0.29,  0.05],
                     [-0.11, -0.29],
                     [ 0.05,  0.29],
                     [-0.03, -0.05]], requires_grad=True)
w = w.transpose(0,1)

print(x.size(),x.dtype)
print(w.size(),w.dtype)
#torch.Size([1, 2, 2]) torch.float32
#torch.Size([2, 4]) torch.float32

o1 = torch.matmul(x1,w)
o2 = torch.matmul(x2,w)

print(o1)
print(o2)
# tensor([[[-0.02738600,  0.22973600, -0.23244201,  0.03911200],
#          [-0.12149499,  0.18984099, -0.17019099,  0.03634500]]],
#        grad_fn=<UnsafeViewBackward0>)
print(torch.equal(o1,o2[:,:-1,:]))
#False
```
seems like a loss of precision?

### Versions

could reproduce on cpu and gpu of different version, seems not a env problems

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by malfet** [comment]
Can not  repro your minimal example neither  on my M1 machine nor in [Colab](https://colab.research.google.com/drive/1ODGmw8Q9HDBWS5m6dVYqM0NJp0IH1i20?usp=sharing), so please share some details about the system you are running it on (by running collect_env script for example)

Also, please note, that floating point operations are imprecise by nature, and changing order of operations can indeed change the result. Do you mind printing `(o1-o2[:,:-1,:]).abs().max()` instead of just equals?

**Comment by LamForest** [comment]
I cannot reproduce on main either

**Comment by Chandler-Bing** [comment]
@malfet  hi, you can repro it using follow code (all tensors are randomly generated),  [ptrblck's](https://discuss.pytorch.org/t/torch-nn-linear-bug/199684/4?u=chandler-bing) answer suggests it seems like a normal behavior. so I think it maybe  not a bug
```
import torch
torch.set_printoptions(precision=8)
x = torch.tensor([[[ 0.74851763,  0.53351247, -0.62923586,  0.02778959],
                [ 0.92442727,  0.91173172,  0.41091037,  0.34944904]]]
                 )
x1 = x[:,:-1,:]
x2 = x
w = torch.tensor([[-0.17140836,  0.27952832,  0.21300077,  0.16653723],
                  [ 0.15126991, -0.02109182,  0.41659349,  0.20512718],
                  [-0.30581605, -0.35315818, -0.26448679, -0.11028063]],
                 requires_grad=True)
w = w.transpose(0,1)

print(x.size(),x.dtype)
print(w.size(),w.dtype)
#torch.Size([1, 2, 2]) torch.float32
#torch.Size([2, 4]) torch.float32

o1 = torch.matmul(x1,w)
o2 = torch.matmul(x2,w)

print(o1)
print(o2)

print(torch.equal(o1,o2[:,:-1,:]))
#False
```
![image](https://github.com/pytorch/pytorch/assets/29994840/c5f949bc-57ae-4cac-afcb-7dc932f044f2)

also, my env by running collect_env script
```
C:\Users\my\Anaconda3\python.exe D:/projects/clean/src/xx.py
Collecting environment information...
PyTorch version: 2.1.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Revision=

Versions of relevant libraries:
[pip3] flake8==3.9.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.1
[pip3] numpydoc==1.1.0
[pip3] torch==2.1.1
[pip3] torchdata==0.6.1
[pip3] torchvision==0.15.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           haa95532_296  
[conda] mkl-service               2.3.0            py38h2bbff1b_1  
[conda] mkl_fft                   1.3.0            py38h277e83a_2  
[conda] mkl_random                1.2.1            py38hf11a4ad_2  
[conda] numpy                     1.22.1                   pypi_0    pypi
[conda] numpy-base                1.20.1           py38haf7ebc8_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     2.1.1                    pypi_0    pypi
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
```

**Comment by albanD** [comment]
From the print your shared, these are equal up to floating point precision. And the discrepancy is expected as described in details here: https://pytorch.org/docs/stable/notes/numerical_accuracy.html#batched-computations-or-slice-computations

Closing as expected behavior but feel free to ask for any clarification!
